README
README 3/3/91

What it is:
  This distribution contains the files README, interface.c, interface.h,
  and sample.c. The files interface.c and interface.h contain definitions
  and declarations for some generic Internet-interfacing code. The best
  form of documentation is by example, in my opinion, so I included
  sample.c, which is a simple echo-mud that supports WHO, QUIT, and @shutdown
  as commands. Read it over a few times to get the gist of how to use
  the interface package.

  Here's some sample pseudocode sorta stuff, demonstrating the general
  procedure your program should follow:

    init_port(My Port Number, My Maximum Buffer Size);
    Do Our Own Start Up Stuff;
    while (We Still Want To Run) {
      Connection=get_event();
      switch (conn_status(Connection)) {
      case INPUT_RECEIVED:
        Handle the Input, using read_conn();
        break;
      case NEW_CONNECTION:
        Do Whatever Stuff We Need To Do With A New Connection;
        break;
      case CONNECTION_DEAD:
        Do Our Own Clean Up;
        shutdown_conn(Connection);
        break;
      }
    }
    Do Some More Of Our Own Clean Up;
    shutdown_port();



Configuring interface.c and interface.h:
  The following three values are #define'd in interface.h:

  MAX_CONNS         58    Maximum number of connections we'll take at a time
  MAX_INBUF         96    Maximum number of input lines to buffer
  MAX_LINE_PER_SEC  10    Maximum number of lines per second we'll receive
                          before the connection gets closed by the anti-
                          spammer (unless the connection has been marked
                          SPAM_OK with connctl())


Functions Declared In interface.h:

int init_port(int portno, int bufsiz);
  This connects a socket to portno, and configures it to receives a maximum
  string length of bufsiz-1. Returns 0 on failure.

void shutdown_port(void);
  This closes the port and its associated connections, and does some
  routine clean-up. Should be done before the program exits.

void shutdown_conn(CONN cd);
  This closes the connection and empties the input buffer.

int write_conn(CONN cd, char *msg);
  This writes the string msg to the specified connection. It returns -1
  on error.

char *read_conn(CONN cd);
  This returns the next string in the input queue of the connection. It
  returns NULL if the connection has no input queued.

CONN get_event(void);
  This waits for some activity on one of the open connections, and will
  return the connection for which the activity has occurred. The program
  calling get_event() should call conn_status() on the connection to
  determine the nature of the event.

int connctl(CONN cd, int oper, int data);
  This sets certain modes of the connection. Currently, the only
  operations permitted are:
    CONN_SET_FLAGS           This sets the flags on the connection.
                             The only flag supported currently is
                             SPAM_OK, which turns off the anti-spammer
                             for that connection when the flag is set.
    CONN_GET_FLAGS           Returns the flags set on the connection.
                             Data can be anything.
  connctl() returns -1 on error.

int conn_status(CONN cd);
  This returns the status of the connection. This will be one of the
  following:
    INPUT_RECEIVED           There is input queued on this connection.
                             Use read_conn() to retrieve it.
    NEW_CONNECTION           This is a new connection that has just been
                             formed.
    CONNECTION_DEAD          The connection has died on the client's end,
                             or the anti-spammer code has detected spamming.
                             The program should shutdown_conn() on this
                             connection.

CONN get_conn(int refno);
  This takes a number from 0 to MAX_CONNS-1 and indexes into the connection
  table. It will return a CONN if the connection is open, or NULL if the
  connection is closed or if the refno is out of bounds.

struct sockaddr_in get_inet_addr(CONN cd);
  This function returns the address of the connection.

int get_fd(CONN cd);
  This function returns the file descriptor associated with the connection.





interface.c and interface.h are 100% public domain. Do whatever the hell
you want with them.

DISCLAIMER: The author accepts no responsibility or liability for anything
  this code may do when compiled or run. Cope.

NOTE: The author doesn't care whether you think the code is pretty or not.
      The author doesn't believe in commenting.
      If there's a bug, email me about it. If you just want to whine about
      how gross it is, don't.

The author is:       Patrick J Wetmore
                     floyd@rpi.edu

CHANGES
CHANGES 4/29/91:

Calls to shutdown() now include the second argument, a 2.
A useless call to fcntl() was removed.
The call to accept() in make_new_conn was changed so that the 3rd
parameter passed contains a pointer to the size of the structure,
not a 0.

In sample.c: (of little importance)
A lot of people couldn't compile it because they had non-backward-
compatible /usr/include/inet/in.h's. I changed the address-printing
bit of code for the WHO command to a call to inet_ntoa

None of the above should change your code even a little bit. Just re-compile
with the new interface.c.

Patrick J. Wetmore
floyd@rpi.edu


sample.c
#include <stdio.h>
#include "interface.h"

extern void *malloc();
extern char *inet_ntoa();

struct cdlist {
  CONN cd;
  struct cdlist *next;
};

void notify_all(struct cdlist *cds, char *msg)
{
  while (cds) {
    write_conn(cds->cd,msg);
    write_conn(cds->cd,"\n");
    cds=cds->next;
  }
}

main() {
  char ourbuf[512],*theirbuf;
  CONN curr_cd;
  struct cdlist *cds,*tmp,*prev;

  cds=NULL;
  if (!init_port(3939,512)) {
    fprintf(stderr,"Failed to open port 3939\n");
    exit(-1);
  }
  while (1) {
    curr_cd=get_event();
    switch (conn_status(curr_cd)) {
    case INPUT_RECEIVED:
      theirbuf=read_conn(curr_cd);
      if (!strcmp(theirbuf,"WHO")) {
	tmp=cds;
	sprintf(ourbuf,"FD\tAddress\n");
	write_conn(curr_cd,ourbuf);
	while (tmp) {
	  sprintf(ourbuf,"%d\t%s\n",
		  (int) get_fd(tmp->cd),inet_ntoa(get_inet_addr(tmp->cd).
						  sin_addr));
	  write_conn(curr_cd,ourbuf);
	  tmp=tmp->next;
	}
	break;
      } else if (!strcmp(theirbuf,"QUIT")) {
	write_conn(curr_cd,"Ciao babe.\n");
	if (cds->cd==curr_cd) {
          tmp=cds;
	  cds=cds->next;
	  free(tmp);
	} else {
	  prev=cds;
	  tmp=prev->next;
	  while (tmp) {
	    if (tmp->cd==curr_cd) {
	      prev->next=tmp->next;
	      free(tmp);
	      break;
	    }
	    prev=tmp;
	    tmp=tmp->next;
	  }
	}
	shutdown_conn(curr_cd);
	break;
      } else if (!strcmp(theirbuf,"@shutdown")) {
	notify_all(cds,"Shutting down, babes.\n");
	shutdown_port();
	exit(0);
      } else
        notify_all(cds,theirbuf);
      break;
    case NEW_CONNECTION:
      tmp=malloc(sizeof(struct cdlist));
      tmp->cd=curr_cd;
      tmp->next=cds;
      cds=tmp;
      write_conn(curr_cd,"Howdy!\n");
      break;
    case CONNECTION_DEAD:
      if (cds->cd==curr_cd) {
	tmp=cds->next;
	shutdown_conn(curr_cd);
	free(cds);
	cds=tmp;
      } else {
	prev=cds;
	tmp=cds->next;
	while (tmp) {
	  if (tmp->cd==curr_cd) {
	    prev->next=tmp->next;
	    shutdown_conn(curr_cd);
	    free(tmp);
	    break;
	  }
	  prev=tmp;
	  tmp=tmp->next;
	}
      }
      break;
    }
  }
}


interface.c
#define __INTERFACE

#include <stdio.h>
#include <time.h>
#include <ctype.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <fcntl.h>
#include "interface.h"

#define difftime(x,y) (x-y)
#define NUM_FM 16

static int ibufsiz;
static int sock_fd;
static int fd_lim;
static char *internal_buf;
static struct conn connlist[MAX_CONNS];
static struct {
    int avail;
    char *buffer;
  } fmbuf[NUM_FM];

static char *fastmalloc(int size)
{
  int loop=0;

  if (size>ibufsiz)
    return (char *) malloc(size);
  while (loop<NUM_FM) {
    if (fmbuf[loop].avail) {
      fmbuf[loop].avail=0;
      return fmbuf[loop].buffer;
    }
    loop++;
  }
  return (char *) malloc(size);
}

static void fastfree(char *buf)
{
  int loop=0;

  while (loop<NUM_FM) {
    if (fmbuf[loop].buffer==buf) {
      fmbuf[loop].avail=1;
      return;
    }
    loop++;
  }
  free(buf);
}

CONN get_conn(int refno)
{
  if (refno<0 || refno>=MAX_CONNS) return NULL;
  return (connlist[refno].is_free) ? NULL : &(connlist[refno]);
}

int connctl(CONN cd, int oper, int data)
{
  switch (oper) {
  case CONN_SET_FLAGS:
    cd->flags=data;
    return 1;
  case CONN_GET_FLAGS:
    return cd->flags;
  default:
    return -1;
  }
}

int init_port(int portno, int bufsiz)
{
  int opt=1;
  struct sockaddr_in server;
  int loop;

  sock_fd=socket(AF_INET,SOCK_STREAM,0);
  if (sock_fd<0) return 0;
  server.sin_family=AF_INET;
  server.sin_addr.s_addr=INADDR_ANY;
  server.sin_port=htons(portno);
  if (bind(sock_fd,(struct sockaddr *) &server, sizeof(server)))
    return 0;
  listen(sock_fd,5);
  fd_lim=sock_fd+1;
  ibufsiz=bufsiz;
  for (opt=0;opt<MAX_CONNS;opt++) {
    connlist[opt].is_free=1;
    connlist[opt].partial=(char *) malloc(ibufsiz);
    *(connlist[opt].partial)='\0';
  }
  internal_buf=(char *) malloc(ibufsiz);
  while (loop<NUM_FM) {
    fmbuf[loop].avail=1;
    fmbuf[loop].buffer=(char *) malloc(ibufsiz);
    loop++;
  }
  return 1;
}

void shutdown_conn(CONN cd)
{
  if (cd->is_free) return;
  cd->is_free=1;
  while (cd->inbuf_count) free(cd->inbuf[(cd->inbuf_count)--]);
  shutdown(cd->fd,2);
  close(cd->fd);
  *(cd->partial)='\0';
}

void shutdown_port(void)
{
  int loop;

  for (loop=0;loop<MAX_CONNS;loop++) {
    shutdown_conn(&(connlist[loop]));
    free(connlist[loop].partial);
  }
  close(sock_fd);
  free(internal_buf);
  for (loop=0;loop<NUM_FM;loop++) free(fmbuf[loop].buffer);
}

char *read_conn(CONN cd)
{
  int loop=0;

  if (!(cd->inbuf_count)) return NULL;
  strcpy(internal_buf,cd->inbuf[0]);
  fastfree(cd->inbuf[0]);
  while (loop<cd->inbuf_count) cd->inbuf[loop]=cd->inbuf[++loop];
  --(cd->inbuf_count);
  return internal_buf;
}

CONN make_new_conn(void)
{
  int new_fd,new_cd=(-1),size=sizeof(struct sockaddr_in),loop=0;
  struct sockaddr_in tmpaddr;

  loop=sizeof(struct sockaddr);
  new_fd=accept(sock_fd,(struct sockaddr *) &tmpaddr,
		&size);
  if (new_fd<0) return NULL;
  while (loop<MAX_CONNS)
    if (connlist[loop++].is_free) {
      new_cd=loop-1;
      break;
    }
  if (new_cd==-1) {
    shutdown(new_fd,2);
    close(new_fd);
    return NULL;
  }
  if (new_fd>=fd_lim) fd_lim=new_fd+1;
  connlist[new_cd].fd=new_fd;
  connlist[new_cd].is_free=0;
  connlist[new_cd].address=tmpaddr;
  connlist[new_cd].inbuf_count=0;
  connlist[new_cd].flags=0;
  connlist[new_cd].last_input_time=time(NULL);
  connlist[new_cd].input_this_sec=0;
  return &(connlist[new_cd]);
}

static int add_line(CONN cd,char *buf,int pos,int end,time_t now)
{
  int count=0;
  char *tmp;

  if ((difftime(now,cd->last_input_time)<=1) &&
      ((++(cd->input_this_sec))>MAX_LINE_PER_SEC) &&
      (!(cd->flags & SPAM_OK)))
    return 0;
  if (cd->inbuf_count==MAX_INBUF) return 1;
  if (difftime(now,cd->last_input_time)>1)
    cd->input_this_sec=0;
  cd->last_input_time=now;
  tmp=(char *) fastmalloc(end-pos+1);
  while (pos<end) {
    if (isgraph(buf[pos]) || (buf[pos]==' ')) tmp[count++]=buf[pos];
    pos++;
  }
  tmp[count]='\0';
  cd->inbuf[(cd->inbuf_count)++]=tmp;
  return 1;
}
  

static int process_input(CONN cd)
{
  int len,pos=0,start=0,partial_len;
  time_t now;

  partial_len=strlen(cd->partial);
  if (partial_len>=ibufsiz-1)
    partial_len=0;
  strcpy(internal_buf,cd->partial);
  *(cd->partial)='\0';
  len=read(cd->fd,internal_buf+partial_len,ibufsiz-1-partial_len);
  if (len<=0) return 0;
  len+=partial_len;
  now=time(NULL);
  while (pos<len) {
    if (internal_buf[pos]=='\n') {
      if (!add_line(cd,internal_buf,start,pos,now)) return 0;
      start=pos+1;
    }
    pos++;
  }
  if (pos-start) {
    strncpy(cd->partial,internal_buf+start,pos-start);    
    *(cd->partial+pos-start+1)='\0';
  }
  return 1;
}
    

CONN get_event(void)
{
  fd_set input_set;
  int loop=0;
  CONN new_conn;

  while (1) {
    loop=0;
    while (loop<MAX_CONNS) {
      if ((!(connlist[loop].is_free)) && (connlist[loop].inbuf_count)) {
        connlist[loop].status=INPUT_RECEIVED;
        return &(connlist[loop]);
      }
      loop++;
    }
    loop=0;
    FD_ZERO(&input_set);
    while (loop<MAX_CONNS) {
      if (!(connlist[loop].is_free)) {
        FD_SET(connlist[loop].fd,&input_set);
      }
      loop++;
    }
    FD_SET(sock_fd,&input_set);
    select(fd_lim,&input_set,NULL,NULL,NULL);
    if (FD_ISSET(sock_fd,&input_set)) {
      new_conn=make_new_conn();    
      if (new_conn) {
        new_conn->status=NEW_CONNECTION;
        return new_conn;
      }
    }
    loop=0;
    while (loop<MAX_CONNS) {
      if ((!(connlist[loop].is_free)) && (FD_ISSET(connlist[loop].fd,
						   &input_set)))
        if (!process_input(&(connlist[loop]))) {
          connlist[loop].status=CONNECTION_DEAD;
          return &(connlist[loop]);
	} else {
          connlist[loop].status=INPUT_RECEIVED;
	}
      loop++;
    }
  }
}


interface.h
#define MAX_CONNS 58
#define MAX_INBUF 96
#define MAX_LINE_PER_SEC 10

#ifndef __INTERFACE
#include <sys/types.h>
#include <netinet/in.h>
#endif /* __INTERFACE */

struct conn {
  int fd;
  int is_free;
  struct sockaddr_in address;
  char *inbuf[MAX_INBUF];
  int inbuf_count;
  char *partial;
  int status;
  int flags;
  time_t last_input_time;
  int input_this_sec;
};

typedef struct conn *CONN;

int init_port(int portno, int bufsiz);
void shutdown_port(void);
void shutdown_conn(CONN cd);
#define write_conn(DESCRIPTOR,STRING) \
  write(DESCRIPTOR->fd,STRING,strlen(STRING))
char *read_conn(CONN cd);
CONN get_event(void);
int connctl(CONN cd, int oper, int data);
#define conn_status(DESCRIPTOR) \
  (DESCRIPTOR->status)
CONN get_conn(int refno);
#define get_inet_addr(X) \
  (X->address)
#define get_fd(X) \
  (X->fd) 

/* possible statuses for conn_status to return */
#define INPUT_RECEIVED 0
#define NEW_CONNECTION 1
#define CONNECTION_DEAD 2

/* possible operations with connctl */
#define CONN_SET_FLAGS 1
#define CONN_GET_FLAGS 2

/* possible flags that can be set on the CONN with connctl */
#define SPAM_OK 1


